src/main/java/edu/harvard/iq/dataverse/api/DownloadInstanceWriter.java

```diff
@@ -403,9 +427,14 @@ public void writeTo(DownloadInstance di, Class<?> clazz, Type type, Annotation[]
                         long contentSize;
                         boolean useChunkedTransfer = false;
                         //if ((contentSize = getFileSize(di, storageIO.getVarHeader())) > 0) {
-                        if ((contentSize = getContentSize(storageIO)) > 0) {
+                        if ((contentSize = getContentSize(storageIO)) > 0 && ranges.isEmpty()) {
                             logger.fine("Content size (retrieved from the AccessObject): " + contentSize);
                             httpHeaders.add("Content-Length", contentSize);
+                        } else if (!ranges.isEmpty()) {
+                            // For now we only support a single range.
+                            long rangeContentSize = ranges.get(0).getLength();
+                            logger.fine("Content size (Range header in use): " + rangeContentSize);
+                            httpHeaders.add("Content-Length", rangeContentSize);
                         } else {
```

Comment:
Note this case in the original code ("do nothing, if content size is not known").
This is under the assumption that there are some download cases (subsetting, specifically, and maybe others) where the final stream we are sending to the client is being generated dynamically, on the fly. In that case, `getContentSize(storageIO)` will return -1. So, if we know the size of the stream we are about to serve, we add this explicit `Content-Length` header. Otherwise, we don't - and then Glassfish/Payara automatically switches to the "chunked encoding" mode. (this is why the lines like `//httpHeaders.add("Transfer-encoding", "chunked");` are commented out - we no longer have to do it in our code ourselves). 
In the current PR, the code further below is written under the assumption that we will always know the total size of the download (and thus will know the value of `leftToRead`). It needs to be changed back to where this is not necessarily the case. (So ideally the `leftToRead` value should *only* be used when it's a Range request). And we should probably throw an exception when a Range is requested, but the final size is not known/i.e., the output is a dynamic stream.

Reply:
As part of the refactoring in 29ea256 I'm now preserving the old logic for non-range downloads. Only range downloads have this `leftToRead` concept now.